tests/test_cli.py

```diff
@@ -0,0 +1,140 @@
+        'buku: error: unrecognized arguments: --unknown',
+    ]
+
+@pytest.mark.parametrize('argv', [['--help'], ['foo', 'bar', '--help']])
+def test_help(BukuDb, exit, piped_input, argv):
+    with mock.patch('buku.ExtendedArgumentParser.print_help') as print_help:
+        with pytest.raises(SystemExit):
+            buku.main(argv)
+    BukuDb.assert_not_called()
+    print_help.assert_called_with()
+    exit.assert_called_with(0)
+
+@pytest.mark.parametrize('argv', [[], ['--nostdin']])
+def test_prompt(BukuDb, bdb, piped_input, prompt, argv):
+    with pytest.raises(SystemExit):
+        buku.main(argv)
+    piped_input.assert_not_called()
+    BukuDb.assert_called_with()
+    prompt.assert_called_with(bdb, None)
+    bdb.close_quit.assert_called_with(0)
+
+
+@pytest.mark.parametrize('fetch_params', [
+    {'offline': True},
+    {'url_redirect': True},
+    {'tag_redirect': True, 'tag_error': True},
+    {'url_redirect': True, 'tag_redirect': 'redirect', 'tag_error': 'error'},
+    {'url_redirect': True, 'tag_redirect': 'redirect', 'del_range': [], 'del_error': range(400, 600)},
+    {'url_redirect': True, 'tag_redirect': 'redirect', 'tag_error': 'error',
+     'del_range': ['400-404', '500'], 'del_error': {400, 401, 402, 403, 404, 500}},
+])
+@pytest.mark.parametrize('value_params', [
+    {'add_tags': ['foo,bar', 'baz'], 'tags_fetch': False, 'tags_in': ',bar baz,foo,', 'title': ''},
+    {'tag': ['foo', 'bar,baz'], 'tags_fetch': False, 'tags_in': ',baz,foo bar,', 'title': 'Custom Title'},
+    {'add_tags': ['+', 'foo', 'bar', 'baz'], 'tags_in': ',foo bar baz,', 'comment': ''},
+    {'tag': ['+', 'foo,bar,baz'], 'tags_in': ',bar,baz,foo,', 'comment': 'Custom Description'},
+    {'add_tags': ['-', 'foo', 'baz', 'baz'], 'tags_except': ',foo baz baz,', 'immutable': False},
+    {'tag': ['-', 'foo,', ',baz,', ',baz'], 'tags_except': ',baz,foo,', 'immutable': True},
+    {'add_tags': ['foo,baz,bar'], 'tag': ['baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['+', 'foo,baz,bar'], 'tag': ['baz,', 'qux,'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['foo,baz,', 'bar,'], 'tag': ['+', 'baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['-', 'foo,baz,bar,'], 'tag': ['baz,', 'qux,'],
+     'tags_fetch': False, 'tags_in': ',baz,qux,', 'tags_except': ',bar,baz,foo,'},
+    {'add_tags': ['foo,baz,', 'bar,'], 'tag': ['-', 'baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,', 'tags_except': ',baz,qux,'},
+    {'add_tags': ['-', 'foo,baz,bar,'], 'tag': ['-', 'baz,', 'qux,'], 'tags_except': ',bar,baz,foo,qux,'},
+    {'add_tags': ['-', 'foo,baz,', 'bar,'], 'tag': ['+', 'baz,', 'qux,'],
+     'tags_in': ',baz,qux,', 'tags_except': ',bar,baz,foo,'},
+    {'add_tags': ['+', 'foo,baz,', 'bar,'], 'tag': ['-', 'baz,', 'qux,'],
+     'tags_in': ',bar,baz,foo,', 'tags_except': ',baz,qux,'},
+])
```

Comment:
That's 84 test cases… which is still way better than ~3000 that I got when I tried out all combinations (…and I wasn't even done writing them :laughing:)